### PR TITLE
mgr/rgwam: use realm/zonegroup/zone method arguments for period update

### PR DESCRIPTION
**Pull request by aaSharma14**
period update method currently uses default zonegroup/zone for period update commit. This PR is to use realm/zg/zone provided in the arguments for period update. Also modify the `set-rgw-credentials`command to save the initial "default" zone's access creds in the dashboard config even if a realm is created in the setup
 
Fixes: https://tracker.ceph.com/issues/67897

Signed-off-by: Aashish Sharma <aasharma@redhat.com>

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "quincy"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

- When filling out the below checklist, you may click boxes directly in the GitHub web UI.  When entering or editing the entire PR message in the GitHub web UI editor, you may also select a checklist item by adding an `x` between the brackets: `[x]`.  Spaces and capitalization matter when checking off items this way.

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [ ] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [ ] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
- `jenkins test rook e2e`
</details>

### CODE REVIEW

File: src/python-common/ceph/rgw/rgwam_core.py
Comment (cbodley) [inline src/python-common/ceph/rgw/rgwam_core.py:831]: why is this part necessary?

Comment (aaSharma14) [reply src/python-common/ceph/rgw/rgwam_core.py:831]: This is mentioned in the docs - https://docs.ceph.com/en/quincy/radosgw/multisite/#pulling-the-realm-configuration. Moreover as per the current implementation in dashboard, When we do realm pull on a cluster with a rgw service running on the initial `default` named zonegroup and zone, we store the rgw credentials for that realm in the dashboard config like - `{"realm1": "PFIR20LCC8KP126OB7ON"}`. Also we restart the rgw daemons in this process, so everytime we visit the buckets/users page it fetches the daemon credentials from this config using the realm name. If we do not make the first realm default, the existing rgw daemons running on `default` named zonegorup and zone are not able to fetch the access/secret keys from the config and throw 404 Credentials not found error because they are not included in any realm. If we make this realm default,on daemon restart the existing daemons running on `default` zg/zone get deployed under this realm and we can get credentials from the config.
If we already have realm/s in place and all daemons are running on any of the realm, then we do not see this issue.

Comment (cbodley) [reply src/python-common/ceph/rgw/rgwam_core.py:831]: sorry, i'm not following the requirement to store things in the default zone/zonegroup or restart existing services

Comment (aaSharma14) [reply src/python-common/ceph/rgw/rgwam_core.py:831]: In short, currently we do not allow a service to be running on the default zone/zonegroup if a realm is created in the setup (because we no longer store the access/secret keys of the user used to create that service), So we restart that service to make it run on newly created realm/zg/one by making the pulled realm default if it's the first one.

Comment (cbodley) [reply src/python-common/ceph/rgw/rgwam_core.py:831]: > because we no longer store the access/secret keys of the user used to create that service

can you clarify this part? are we talking about the access/secret keys stored in the zone for rgw multisite? or the creds that the dashboard uses to access the rgw admin apis?

Comment (aaSharma14) [reply src/python-common/ceph/rgw/rgwam_core.py:831]: The access/secret keys stored in the zone. We store these keys in a dashboard config like  - {"realm1": "PFIR20LCC8KP126OB7ON"}. So a rgw service running on default zg/zone doesn't have these keys stored in the config so it's unable to authenticate. This is all existing functionality. If we do not want to make the first realm as default then we need to change the way we store the zone keys in config and store the "default" zone's keys as well so the existing services running on this "default" zone are able to fetch these keys from the config.

Comment (aaSharma14) [reply src/python-common/ceph/rgw/rgwam_core.py:831]: Till there is no realm, the "default" zone's keys are stored as plain strings (and not as a dict like I mentioned above which uses realm name to get the corresponding keys). When a realm is created we expect every service to be deployed on a realm so we can use service's realm name to fetch the keys from the dict

Comment (cbodley) [reply src/python-common/ceph/rgw/rgwam_core.py:831]: > The access/secret keys stored in the zone. We store these keys in a dashboard config like - {"realm1": "PFIR20LCC8KP126OB7ON"}. So a rgw service running on default zg/zone doesn't have these keys stored in the config so it's unable to authenticate.

this makes it sound like you're referring to the "dashboard" user's credentials created by `ceph dashboard set-rgw-credentials`, which the dashboard uses to authenticate against rgw's admin apis. isn't that already configured and working for the default zone/zonegroup before we create any realms?

Comment (aaSharma14) [reply src/python-common/ceph/rgw/rgwam_core.py:831]: Yes, the one's we get from `ceph dashboard set-rgw-credentials`. It works till there is no realm because we store the access/secret keys as plain strings because there are just one set of values to be used. When there is a realm/s we store the keys in a dict with realm name as keys and we donot store the "default" zone's keys in the config anymore because it has no realm and therefore it fails to authenticate

Comment (cbodley) [reply src/python-common/ceph/rgw/rgwam_core.py:831]: if the admin has a default zone/zonegroup with no realm, they'll use commands like `radosgw-admin user create`, `radosgw-admin bucket rm` etc to operate on it. if we create a realm and set it as default, those commands will operate inside that realm instead, and won't do what the admin expects

the dashboard is meant to manage several independent rgw services. to maintain that independence, it shouldn't change or rely on any defaults

rgw intentionally changed that behavior in https://github.com/ceph/ceph/pull/59422/commits/d64e273de1623fe784a3b8c9d2c21990f78b7be2, so i'd prefer not to add it back here

**Review by adk3798 — COMMENTED** [review]
I've also seen a failure if you try to run `ceph rgw realm bootstrap...` twice on the same cluster

```
stderr: 2024-08-27T19:37:51.168+0000 7fd1a6327800  0 period (6d38bcef-5c6a-4f55-8f30-9e7e9f64b13f does not have zone e06d414e-da89-4495-af09-3d96b752c3a4 configured
Sending period to new master zone 6f4c1a3f-b27a-4978-a036-2688c761a317
2024-08-27T19:37:51.501+0000 7fd1a6327800  0 ERROR: endpoints not configured for upstream zone
request failed: (22) Invalid argument
failed to commit period: (22) Invalid argument
```

did you also look at / test that case by any chance? @aaSharma14

File: src/python-common/ceph/rgw/rgwam_core.py
Comment (aaSharma14) [reply src/python-common/ceph/rgw/rgwam_core.py]: Thanks @cbodley , updated the PR again

**Comment by aaSharma14** [comment]
> I've also seen a failure if you try to run `ceph rgw realm bootstrap...` twice on the same cluster
> 
> ```
> stderr: 2024-08-27T19:37:51.168+0000 7fd1a6327800  0 period (6d38bcef-5c6a-4f55-8f30-9e7e9f64b13f does not have zone e06d414e-da89-4495-af09-3d96b752c3a4 configured
> Sending period to new master zone 6f4c1a3f-b27a-4978-a036-2688c761a317
> 2024-08-27T19:37:51.501+0000 7fd1a6327800  0 ERROR: endpoints not configured for upstream zone
> request failed: (22) Invalid argument
> failed to commit period: (22) Invalid argument
> ```
> 
> did you also look at / test that case by any chance? @aaSharma14

@adk3798 , I haven't tried this scenario. Is this something being caused by this PR or was this error there before this PR as well?

**Comment by aaSharma14** [comment]
> I've also seen a failure if you try to run `ceph rgw realm bootstrap...` twice on the same cluster
> 
> ```
> stderr: 2024-08-27T19:37:51.168+0000 7fd1a6327800  0 period (6d38bcef-5c6a-4f55-8f30-9e7e9f64b13f does not have zone e06d414e-da89-4495-af09-3d96b752c3a4 configured
> Sending period to new master zone 6f4c1a3f-b27a-4978-a036-2688c761a317
> 2024-08-27T19:37:51.501+0000 7fd1a6327800  0 ERROR: endpoints not configured for upstream zone
> request failed: (22) Invalid argument
> failed to commit period: (22) Invalid argument
> ```
> 
> did you also look at / test that case by any chance? @aaSharma14

@adk3798 , Can you please retry this one top of this PR? i didn't face any issue running this command twice on same cluster

**Review by cbodley — APPROVED** [review]
the addition of the `zone` argument to `update_period()` looks good. i see several other changes that aren't described in the commit message, so i don't understand why they were necessary. but approving regardless

**Comment by adk3798** [comment]
> > I've also seen a failure if you try to run `ceph rgw realm bootstrap...` twice on the same cluster
> > ```
> > stderr: 2024-08-27T19:37:51.168+0000 7fd1a6327800  0 period (6d38bcef-5c6a-4f55-8f30-9e7e9f64b13f does not have zone e06d414e-da89-4495-af09-3d96b752c3a4 configured
> > Sending period to new master zone 6f4c1a3f-b27a-4978-a036-2688c761a317
> > 2024-08-27T19:37:51.501+0000 7fd1a6327800  0 ERROR: endpoints not configured for upstream zone
> > request failed: (22) Invalid argument
> > failed to commit period: (22) Invalid argument
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > did you also look at / test that case by any chance? @aaSharma14
> 
> @adk3798 , Can you please retry this one top of this PR? i didn't face any issue running this command twice on same cluster

It seems like with your changes this now works, thanks

**Review by adk3798 — APPROVED** [review]
`run-tox-mgr-dashboard-py3` is still failing, but this does seem to have fixed the issue I was seeing with rgw realm bootstrap

**Comment by aaSharma14** [comment]
> `run-tox-mgr-dashboard-py3` is still failing, but this does seem to have fixed the issue I was seeing with rgw realm bootstrap

@adk3798 , I have fixed the test failure, you can merge this PR when the required tests pass

**Comment by adk3798** [comment]
> > `run-tox-mgr-dashboard-py3` is still failing, but this does seem to have fixed the issue I was seeing with rgw realm bootstrap
> 
> @adk3798 , I have fixed the test failure, you can merge this PR when the required tests pass

make check failed with
```
The following tests FAILED:
	 10 - run-tox-mgr-dashboard-lint (Failed)
```
so there might be more to fix?

**Comment by adk3798** [comment]
https://pulpito.ceph.com/adking-2024-09-21_16:57:11-orch:cephadm-wip-adk-testing-2024-09-21-1014-distro-default-smithi/

Failures:
- 6x mds_upgrade_sequence timeouts, known issue
- 3x SELinux denials + stray daemon warnings, known issue
- 2x rgw-ingress timeout, known issue
- test_nfs task failed due to PR in the run
- staggered upgrade with agent enabled failed, known issue
- stray daemon warning in rados_python task (not a big issue, needs an ignorelist entry)

Nothing to block merging outside of the PR causing the test_nfs task failure

**Comment by adk3798** [comment]
jenkins test windows

**Comment by aaSharma14** [comment]
jenkins test windows

**Pull request merged**
